File: virksomhetssertifikat/src/test/java/no/ks/fiks/virksomhetsertifikat/VirksomhetSertifikaterTest.java
package no.ks.fiks.virksomhetsertifikat;

import lombok.NonNull;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import java.net.URL;
import java.util.Collections;

import static org.junit.jupiter.api.Assertions.*;

class VirksomhetSertifikaterTest {

    public static final String CERTIFICATE_ALIAS = "test";
    public static final String KEYSTORE_PASSWORD = "test";
    public static final String PRIVATE_KEY_ALIAS = "test";
    public static final String PRIVATE_KEY_PASSWORD = "test";

    @DisplayName("Laster SIGN sertifikat fra classpath URL")
    @Test
    void signKey() {
        final VirksomhetSertifikaterProperties virksomhetSertifikaterProperties = new VirksomhetSertifikaterProperties();
        final Sertifikat signSertifikat = getSertifikatFraClasspath(SertifikatType.SIGN);
        virksomhetSertifikaterProperties.setSertifikater(Collections.singleton(signSertifikat));
        final VirksomhetSertifikater virksomhetSertifikater = new VirksomhetSertifikater(virksomhetSertifikaterProperties);
        assertNotNull(virksomhetSertifikater.requireSignKeyStore());
        assertThrows(RuntimeException.class, () -> virksomhetSertifikater.requireAuthKeyStore());
        assertThrows(RuntimeException.class, () -> virksomhetSertifikater.requireEncKeyStore());
    }

    @DisplayName("Laster AUTH sertifikat fra classpath URL")
    @Test
    void authKey() {
        final VirksomhetSertifikaterProperties virksomhetSertifikaterProperties = new VirksomhetSertifikaterProperties();
        final Sertifikat signSertifikat = getSertifikatFraClasspath(SertifikatType.AUTH);
        virksomhetSertifikaterProperties.setSertifikater(Collections.singleton(signSertifikat));
        final VirksomhetSertifikater virksomhetSertifikater = new VirksomhetSertifikater(virksomhetSertifikaterProperties);
        assertNotNull(virksomhetSertifikater.requireAuthKeyStore());
        assertThrows(RuntimeException.class, () -> virksomhetSertifikater.requireSignKeyStore());
        assertThrows(RuntimeException.class, () -> virksomhetSertifikater.requireEncKeyStore());
    }

    @DisplayName("Laster ENC sertifikat fra classpath URL")
    @Test
    void encKey() {
        final VirksomhetSertifikaterProperties virksomhetSertifikaterProperties = new VirksomhetSertifikaterProperties();
        final Sertifikat signSertifikat = getSertifikatFraClasspath(SertifikatType.ENC);
        virksomhetSertifikaterProperties.setSertifikater(Collections.singleton(signSertifikat));
        final VirksomhetSertifikater virksomhetSertifikater = new VirksomhetSertifikater(virksomhetSertifikaterProperties);
        assertNotNull(virksomhetSertifikater.requireEncKeyStore());
        assertThrows(RuntimeException.class, () -> virksomhetSertifikater.requireSignKeyStore());
        assertThrows(RuntimeException.class, () -> virksomhetSertifikater.requireAuthKeyStore());
    }

    @DisplayName("Laster SIGN sertifikat fra filsti")
    @Test
    void loadFromWorkingfolder() {
        final URL resource = getClass().getResource("/certs/test.p12");
        assertNotNull(resource);

        final VirksomhetSertifikaterProperties virksomhetSertifikaterProperties = new VirksomhetSertifikaterProperties();
        virksomhetSertifikaterProperties.setSertifikater(Collections.singleton(createSertifikat(resource.getFile(), SertifikatType.SIGN)));
        final VirksomhetSertifikater virksomhetSertifikater = new VirksomhetSertifikater(virksomhetSertifikaterProperties);
        assertNotNull(virksomhetSertifikater.requireSignKeyStore());
        assertThrows(RuntimeException.class, () -> virksomhetSertifikater.requireAuthKeyStore());
        assertThrows(RuntimeException.class, () -> virksomhetSertifikater.requireEncKeyStore());
    }

    private static Sertifikat getSertifikatFraClasspath(SertifikatType sertifikatType) {
        return createSertifikat("classpath:certs/test.p12", sertifikatType);
    }

    private static Sertifikat createSertifikat(@NonNull final String path, @NonNull final SertifikatType sertifikatType) {
        final Sertifikat sertifikat = new Sertifikat();
        sertifikat.setKeystorePath(path);
        sertifikat.setCertificateAlias(CERTIFICATE_ALIAS);
        sertifikat.setKeystorePassword(KEYSTORE_PASSWORD);
        sertifikat.setPrivateKeyAlias(PRIVATE_KEY_ALIAS);
        sertifikat.setPrivateKeyPassword(PRIVATE_KEY_PASSWORD);
        sertifikat.setSertifikatType(sertifikatType);
        return sertifikat;
    }
}

File: virksomhetssertifikat-core/src/test/kotlin/no/ks/fiks/virksomhetsertifikat/VirksomhetSertifikaterTest.kt
package no.ks.fiks.virksomhetsertifikat

import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import java.util.*

internal class VirksomhetSertifikaterTest {
    @DisplayName("Laster SIGN sertifikat fra classpath URL")
    @Test
    fun signKey() {
        val signSertifikat = getSertifikatFraClasspath(SertifikatType.SIGN)
        val virksomhetSertifikater = VirksomhetSertifikater(setOf(signSertifikat))
        Assertions.assertNotNull(virksomhetSertifikater.requireSignKeyStore())
        Assertions.assertThrows(RuntimeException::class.java) { virksomhetSertifikater.requireAuthKeyStore() }
        Assertions.assertThrows(RuntimeException::class.java) { virksomhetSertifikater.requireEncKeyStore() }
    }

    @DisplayName("Laster AUTH sertifikat fra classpath URL")
    @Test
    fun authKey() {
        val authSertifikat = getSertifikatFraClasspath(SertifikatType.AUTH)
        val virksomhetSertifikater = VirksomhetSertifikater(setOf(authSertifikat))
        Assertions.assertNotNull(virksomhetSertifikater.requireAuthKeyStore())
        Assertions.assertThrows(RuntimeException::class.java) { virksomhetSertifikater.requireSignKeyStore() }
        Assertions.assertThrows(RuntimeException::class.java) { virksomhetSertifikater.requireEncKeyStore() }
    }

    @DisplayName("Laster ENC sertifikat fra classpath URL")
    @Test
    fun encKey() {
        val encSertifikat = getSertifikatFraClasspath(SertifikatType.ENC)
        val virksomhetSertifikater = VirksomhetSertifikater(setOf(encSertifikat))
        Assertions.assertNotNull(virksomhetSertifikater.requireEncKeyStore())
        Assertions.assertThrows(RuntimeException::class.java) { virksomhetSertifikater.requireSignKeyStore() }
        Assertions.assertThrows(RuntimeException::class.java) { virksomhetSertifikater.requireAuthKeyStore() }
    }

    @DisplayName("Laster SIGN sertifikat fra filsti")
    @Test
    fun loadFromWorkingfolder() {
        val resource = javaClass.getResource("/certs/test.p12")
        Assertions.assertNotNull(resource)
        val virksomhetSertifikater = VirksomhetSertifikater(setOf(createSertifikat(resource.file, SertifikatType.SIGN)))
        Assertions.assertNotNull(virksomhetSertifikater.requireSignKeyStore())
        Assertions.assertThrows(RuntimeException::class.java) { virksomhetSertifikater.requireAuthKeyStore() }
        Assertions.assertThrows(RuntimeException::class.java) { virksomhetSertifikater.requireEncKeyStore() }
    }

    companion object {
        const val CERTIFICATE_ALIAS = "test"
        const val KEYSTORE_PASSWORD = "test"
        const val PRIVATE_KEY_ALIAS = "test"
        const val PRIVATE_KEY_PASSWORD = "test"
        private fun getSertifikatFraClasspath(sertifikatType: SertifikatType): Sertifikat {
            return createSertifikat("classpath:certs/test.p12", sertifikatType)
        }

        private fun createSertifikat(path: String, sertifikatType: SertifikatType): Sertifikat =
            Sertifikat(
                keystorePath = path,
                certificateAlias = CERTIFICATE_ALIAS,
                keystorePassword = KEYSTORE_PASSWORD,
                privateKeyAlias = PRIVATE_KEY_ALIAS,
                privateKeyPassword = PRIVATE_KEY_PASSWORD,
                sertifikatType = sertifikatType,
            )
    }
}